### ISSUE TITLE
Add support for optional dependencies to asset bundles

### ISSUE THREAD

**Issue by SDKiller** [yiisoft/yii2#7432]
Having spent quite much time trying to arrange `css` files of several bundles in correct order (without success) I finally think that something similar to handling `jsOptions['position']` is nesessary for `css` in `\yii\web\View::registerAssetBundle`, `\yii\web\View::registerCssFile`.

I don't mean  `POS_HEAD`, `POS_READY` like for `js` files, but possibility to define number in `cssOptions['position']` _manually._

Only `depends` for ordering css files is not enough.

I struggled to arrange this to make custom `site.css` be **in the end** - without success:

```
<link rel="stylesheet" href="/assets/281406ee/css/bootstrap.min.css">
<link rel="stylesheet" href="/css/site.css">
<link rel="stylesheet" href="/assets/864382ef/css/bootstrap-editable.css">
<link rel="stylesheet" href="/assets/6d46f80c/css/bootstrap-popover-x.css">
<link rel="stylesheet" href="/assets/7585d6d9/css/bootstrap-typeahead.css">
```

All bundles here depend on bootstrap asset, but even if I place registering widgets bundles in main layout it makes no effect.

And explicit calling `registerAssetFiles` for bundle in main layout results such buggy situation with dependencies:

```
<link rel="stylesheet" href="/assets/864382ef/css/bootstrap-editable.css">
<link rel="stylesheet" href="/assets/281406ee/css/bootstrap.min.css">
```

or

```
<link rel="stylesheet" href="/assets/6d46f80c/css/bootstrap-popover-x.css">
<link rel="stylesheet" href="/assets/281406ee/css/bootstrap.min.css">
```

**Comment by qiangxue** [comment]
What's the problem about letting the bundle containing `site.css` to depend on other bundles?

**Comment by SDKiller** [comment]
The problem is that on some pages we don't need one of the widgets, on some pages - another.
Different layouts are rendered depending on page. And may be in future we'll get rid of some widget.

That is why `AppAsset` containing `site.css` depends on bootstrap which may be considered to be something *constant*.

Anyway, there is something wrong with `css` handling in bundles, maybe it is interfered by `js` ordering and ordering of `css` is done 'as is'.

I finally achieved my goal splitting `AppAsset` into `JsAsset` and `CssAsset`.

And when dependency is not prepended dependent like this:

```
<link rel="stylesheet" href="/assets/864382ef/css/bootstrap-editable.css">
<link rel="stylesheet" href="/assets/281406ee/css/bootstrap.min.css">
```
is a separate issue I suppose.

**Comment by Faryshta** [comment]
the asset is meant to provide a functionality on the frontend. Sometimes it requires css, js, images and even fonts, for example bootstraps. If you only load the JS then you won't have bootstraps at all.

Also many widgets load their own assets under the hood to work seemlessly, `DatePicker` will load the required assets when its called in case you don't have them preloaded in your layout. I am working on angular.js assets and plugins and they will behave the same.

**Comment by SDKiller** [comment]
@Faryshta 

I don't say I load only `css` or `js` - I talk about possibility to arrange `css` files in designed order similar to the feature we have for `js`.

If you are talking about my phrase
> I finally achieved my goal splitting AppAsset into JsAsset and CssAsset.

then - I did not loaded only `js` or `css` - I just split assets and loaded both bundles one by one, but result of that split was desired `css` ordering.

**Comment by samdark** [comment]
CSS files registration takes dependencies into account the same way JavaScript registration does.

**Comment by Faryshta** [comment]
@SDKiller what you want is to load `site.css` before loading the css files from the other assets?

**Comment by samdark** [comment]
So in order for your CSS from asset bundle A to be registered before CSS from asset bundle B you need to make sure that B depends on A.

**Comment by samdark** [comment]
The problem here, of course is that if in some pages you don't need B there should be a way to specify optional dependenct that won't be requested implicitly.

**Comment by twg-group** [comment]
true : need this option to CSS files, like JS. Sometimes need fast include any css...  but without policy of  dependency.

**Comment by diss1dent** [comment]
still need position for css assets in 2017, google pagespeed did'nt like a lot of css in head.. can anyone help?

**Comment by samdark** [comment]
@diss1dent CSS can't be put elsewhere according to HTML spec. What Google suggests is **NOT** putting it into body.

**Comment by dawei101** [comment]
dependencies of assets is not enough for css source

Currently ,  the default order rule of css file loaded is reversed, its means that:
                  **_last register get first position._**

This make the style could not work, if there is some override styles then.

I suggest to make css file first register then go to first place.

**Comment by samdark** [comment]
Won't be implemented in 2.0.
